### PR TITLE
Adopt frame based layout in BottomSheetController

### PR DESCRIPTION
**Pull request by lcapkovic**
### Platforms Impacted
- [x] iOS
- [ ] macOS

#### Problem
**TL;DR: Sizing and positioning the sheet using Auto Layout was causing more problems than it was solving. Directly calculating the frame simplifies the code and sets us up for future planned additions to the bottom sheet.**

The bottom sheet will soon be used in a new experience in our apps which requires a non-rigid mode i.e. the sheet height will change as the user expands the sheet. In preparation for this change, we need more control over how the sheet is sized.

The sheet is currently sized and positioned using constraints, however it turns out this has several problems:
- the height is determined using a complicated set of UILayoutGuides, this is hard to read by new contributors and doesn't extend well to a flexible height mode
- the width sometimes has to fill parent, sometimes not, which adds more constraints
- vertical position is set using a constant,  so we don't benefit from any Auto Layout magic here
- vertical position also depends on the height, so we need to do a dance of waiting for the layout pass to complete and only updating the offset constraint after that, which can trigger another layout pass
- all animations are frame-based (this is a UIKit limitation), so when an animation is interrupted, we need to convert between constraint offsets and the intermediate frame origin
- the sheet sometimes behaves strangely while rotating the screen, it doesn't stay attached to the bottom edge when the animation is in progress
- many more bits that I can't think of ....

So we already do a lot of calculation and barely benefit from Auto Layout's declarative magic.
Manually calculating the frame for any given state is actually fairly simple and we already do most of it anyways. It has many benefits, like a single source of truth for panning / animating / static positioning, simpler code, and easy implementation of the future flexible mode.

#### Solution
- Add a single source of truth for sheet size and position at any given offset - `func sheetFrame(offset:)`.
- Add `headerContentHeight` property which was the last piece missing to let us directly calculate sheet height.
- Add an `.intermediate` sheet expansion state so we are explicit about being in between states.
- In our viewDidLayoutSubviews override, recalculate sheet size and position if we're not in `.intermediate` state. This is a very cheap constant time calculation so we can afford to do it on every layout pass.
- Remove tons of constraints and layout guides that were used for Auto Layout.
- Sheet offset is now just `bottomSheetView.frame.minY` which is an offset from top, so we need to flip a bunch of conditions from the previous constraint-based offset.
- Simplify multiple property change handlers to just call view.setNeedsLayout().

### Verification

- Sanity checks of all sheet features both in `BotomSheetController` and `BottomCommandingController`.
- Tests of animation interruptions, rotating while animating, panning while animating. 
- Interactive sheet hiding.

No visual or functional change, other than fixing a glitch on screen rotation.

### Pull request checklist

This PR has considered:
- [x] Light and Dark appearances
- [x] iOS supported versions (all major versions greater than or equal current target deployment version)
- [x] VoiceOver and Keyboard Accessibility
- [x] Internationalization and Right to Left layouts
- [x] Different resolutions (1x, 2x, 3x)
- [x] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)
- [x] iPad [Pointer interaction](https://developer.apple.com/documentation/uikit/pointer_interactions)
- [ ] [SwiftUI](https://developer.apple.com/tutorials/swiftui) consumption (validation or new demo scenarios needed)
- [x] Objective-C exposure (provide it only if needed)

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/fluentui-apple/pull/855)

### CODE REVIEW

File: ios/FluentUI/Bottom Sheet/BottomSheetController.swift [hunk 1]
```diff
@@ -722,45 +715,42 @@ public class BottomSheetController: UIViewController {
...
-        constraint.priority = .defaultHigh // Lower than required so Auto Layout can enforce max sheet height
-        return constraint
-    }()
-
-    private lazy var sheetWidthConstraint: NSLayoutConstraint = bottomSheetView.widthAnchor.constraint(equalTo: view.widthAnchor)
-
-    private lazy var sheetMaxWidthConstraint: NSLayoutConstraint = bottomSheetView.widthAnchor.constraint(lessThanOrEqualToConstant: Constants.maxSheetWidth)
-
-    private lazy var maxSheetHeightLayoutGuide: UILayoutGuide = UILayoutGuide()
+    // Height of the sheet in its most expanded state
```
Comment (noaMSFT) [inline ios/FluentUI/Bottom Sheet/BottomSheetController.swift:718]: Suggestion: 'fully expanded state' instead of 'most expanded state'

File: ios/FluentUI/Bottom Sheet/BottomSheetController.swift [hunk 2]
```diff
@@ -776,13 +766,13 @@ public class BottomSheetController: UIViewController {
 
         // Minimum padding from top when the sheet is fully expanded
         static let minimumTopExpandedPadding: CGFloat = 25.0
```
Comment (noaMSFT) [inline ios/FluentUI/Bottom Sheet/BottomSheetController.swift:768]: Nit: for the sake of consistency - all CGFloat's should have .0 at the end like this one (25.0) or remove the .0 to be consistent with the rest of the file.

**Review by noaMSFT — APPROVED** [review]
Looks great! Thanks Lukas 👍

File: ios/FluentUI/Bottom Sheet/BottomSheetController.swift [hunk 3]
```diff
@@ -722,45 +715,42 @@ public class BottomSheetController: UIViewController {
...
-    }()
-
-    private lazy var sheetWidthConstraint: NSLayoutConstraint = bottomSheetView.widthAnchor.constraint(equalTo: view.widthAnchor)
-
-    private lazy var sheetMaxWidthConstraint: NSLayoutConstraint = bottomSheetView.widthAnchor.constraint(lessThanOrEqualToConstant: Constants.maxSheetWidth)
-
-    private lazy var maxSheetHeightLayoutGuide: UILayoutGuide = UILayoutGuide()
+    // Height of the sheet in its most expanded state
+    private var expandedSheetHeight: CGFloat {
+        var height: CGFloat = 0
```
Comment (rdeassis) [inline ios/FluentUI/Bottom Sheet/BottomSheetController.swift:720]: I think this can also be a constant, right? (as it's only initialized in the if below and not changed after that)

File: ios/FluentUI/Bottom Sheet/BottomSheetController.swift [hunk 4]
```diff
@@ -626,28 +643,32 @@ public class BottomSheetController: UIViewController {
...
         translationAnimator.pauseAnimation()
         return translationAnimator
     }
 
+    // Vertical offset of bottomSheetView.origin for the given expansion state
+    //
+    // Note: Since .intermediate state doesn't have a well defined offset, this function will
+    // only return the current sheet offset in that case.
     private func offset(for expansionState: BottomSheetExpansionState) -> CGFloat {
         var offset: CGFloat
```
Comment (rdeassis) [inline ios/FluentUI/Bottom Sheet/BottomSheetController.swift:660]: Change offset to let?

File: ios/FluentUI/Bottom Sheet/BottomSheetController.swift [hunk 5]
```diff
@@ -495,25 +497,40 @@ public class BottomSheetController: UIViewController {
...
+    private func sheetFrame(offset: CGFloat) -> CGRect {
+        let availableWidth: CGFloat = view.bounds.width
+        let sheetWidth = shouldAlwaysFillWidth ? availableWidth : min(Constants.maxSheetWidth, availableWidth)
+
+        return CGRect(origin: CGPoint(x: (view.bounds.width - sheetWidth) / 2, y: offset),
+                      size: CGSize(width: sheetWidth, height: expandedSheetHeight))
     }
 
     private func translationRubberBandFactor(for currentOffset: CGFloat) -> CGFloat {
         var offLimitsOffset: CGFloat = 0.0
```
Comment (rdeassis) [inline ios/FluentUI/Bottom Sheet/BottomSheetController.swift:526]: I think we can turn the offLimitsOffset into a constant too:

```
    private func translationRubberBandFactor(for currentOffset: CGFloat) -> CGFloat {
        let offLimitsOffset: CGFloat
        let expandedOffset = offset(for: .expanded)
        let collapsedOffset = offset(for: .collapsed)

        if currentOffset < expandedOffset {
            offLimitsOffset = min(expandedOffset - currentOffset, Constants.maxRubberBandOffset)
        } else if currentOffset > collapsedOffset {
            offLimitsOffset = min(currentOffset - collapsedOffset, Constants.maxRubberBandOffset)
        } else {
            offLimitsOffset = 0
        }

        return max(1.0 - offLimitsOffset / Constants.maxRubberBandOffset, Constants.minRubberBandScaleFactor)
    }
```

Does the change make sense?

File: ios/FluentUI/Bottom Sheet/BottomSheetController.swift [hunk 6]
```diff
@@ -123,38 +134,42 @@ public class BottomSheetController: UIViewController {
...
             }
         }
     }
 
     /// Indicates if the content should be hidden when the sheet is collapsed
     @objc open var shouldHideCollapsedContent: Bool = true {
         didSet {
-            if shouldHideCollapsedContent != oldValue {
-                updateExpandedContentAlpha()
+            if shouldHideCollapsedContent != oldValue && isViewLoaded {
```
Comment (rdeassis) [inline ios/FluentUI/Bottom Sheet/BottomSheetController.swift:151]: nit: use guard/return for these properties? (headerContentHeight, preferredExpandedContentHeight, collapsedContentHeight, shouldHideCollapsedContent, shouldAlwaysFillWidth,collapsedHeightInSafeArea)

**Review by rdeassis — APPROVED** [review]

File: ios/FluentUI/Bottom Sheet/BottomSheetController.swift [hunk 7]
```diff
@@ -36,6 +36,7 @@ public protocol BottomSheetControllerDelegate: AnyObject {
     case expanded // Sheet is fully expanded
     case collapsed // Sheet is collapsed
     case hidden // Sheet is hidden (fully off-screen)
+    case intermediate // Sheet is between states, only used during user interaction / animation
```
[the remark this answers is not in the document]
Comment (lcapkovic) [reply ios/FluentUI/Bottom Sheet/BottomSheetController.swift:39]: I like it

File: ios/FluentUI/Bottom Sheet/BottomSheetController.swift (hunk 4 above)
Comment (lcapkovic) [reply ios/FluentUI/Bottom Sheet/BottomSheetController.swift:660]: I can't believe I didn't know (or forgot) that local let variables don't need to be initialized in line :D
